app/services/providers/copy_to_recruitment_cycle_service.rb

```diff
@@ -44,14 +42,45 @@ def execute(provider:, new_recruitment_cycle:, force: false)
 
   private
 
-    def copy_courses_to_new_provider(provider, new_provider)
+    attr_reader :copy_course_to_provider_service, :copy_site_to_provider_service, :force
+
+    def courses_to_copy(provider, course_codes)
+      courses = []
+      if force
+        if course_codes.nil?
+          Rails.logger.info "no courses will be rollover"
+        else
+          courses = courses_from_course_codes(provider, course_codes)
+        end
+      elsif course_codes.nil?
+        courses = provider.courses
+      else
+        courses = courses_from_course_codes(provider, course_codes)
+      end
+
+      courses
+    end
+
+    def courses_from_course_codes(provider, course_codes)
+      courses = provider.courses.where(course_code: course_codes.to_a.map(&:upcase))
+
+      if courses.count != course_codes.count
+        error_message = "error courses found has discrepancies (#{courses.count} vs #{course_codes.count})"
+        Rails.logger.fatal error_message
+        raise error_message
+      end
+
+      courses
+    end
+
+    def copy_courses_to_new_provider(_provider, new_provider, courses)
       courses_count = 0
 
-      provider.courses.each do |course|
-        new_course = @copy_course_to_provider_service.execute(course: course, new_provider: new_provider)
+      courses.each do |course|
+        new_course = copy_course_to_provider_service.execute(course: course, new_provider: new_provider)
         courses_count += 1 if new_course.present?
-      rescue Exception # rubocop: disable Lint/RescueException
-        logger&.fatal "error trying to copy course #{course.course_code}"
+      rescue StandardError
```

Comment:
I wonder if we should raise the actual error instead of masking it as just something went wrong?

Reply:
I have no idea what sort of exceptions this pathway may encounter

Reply:
It'll likely be an `ActiveRecord::RecordNotSaved` exception since that's what the service is calling `save!` after building the objects. We could just bubble that up? We could leave it for another card but it'll just make trying to reproduce why a particular course wasn't rolledover a bit of a faff.